File: 20112023.py
# ====================== Sum Pair closest to X =======================
class Solution:
    def sumClosest(self, arr, x):
        # code here
        left, right = 0, len(arr) - 1
        closest = float("inf")
        result = [None, None]
        
        while left < right:
            current = arr[left] + arr[right]
            if abs(x - current) < abs(x-closest):
                result = [arr[left], arr[right]]
                closest = current
                
            if current > x:
                right-=1
            else:
                left+=1
        
        # print(result)
        return result
            
# Driver code
if __name__ == '__main__':
    print("Enter total #of test cases... : ")
    tc = int(input())
    while tc > 0:
        print("Enter space seperated  n and x: ")
        n, x = list(map(int, input().strip().split()))
        print("Enter space seperated  values of input array: ")
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.sumClosest(arr, x)
        print("Sum Pair closest to X")
        print(str(ans[0]) + " " + str(ans[1]))
        tc -= 1

File: 20112023.cpp
// ====================== Sum Pair closest to X =======================
#include <iostream>
#include <bits/stdc++.h>
using namespace std;

class Solution{   
public:
    vector<int> sumClosest(vector<int>arr, int x)
    {
        // code here
        int left = 0, right  = arr.size()-1;
        int closest = INT_MAX;
        std::vector<int> result = {0, 0};
        int current = 0;
        
        while (left<right){
            current = arr[left]+arr[right];
            if(std::abs(x-current)<std::abs(x-closest)){
                closest = current;
                result[0] = arr[left];
                result[1] = arr[right];
            }
            
            if(x<current){
                right--;
            }else{
                left++;
            }
        }
        return result;
    }
};

// Driver Code
int main() {
    int t;
    cout<<"Enter the # of test cases ... : ";
    cin >> t;
    while (t--) {
        int n, x;
        cout<<"Enter the values of n and x : ";
        cin >> n >> x;
        vector<int> arr(n);
        cout<<"Enter the values of input array : ";
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        auto ans = ob.sumClosest(arr, x);
        cout<<"Sum Pair closest to X: ";
        cout << ans[0] << " " << ans[1] << "\n";
    }
    return 0;
}

